src/client/chat.js

```diff
@@ -0,0 +1,538 @@
+          if (entry) {
+            ackSet |= 1 << i
+            toSign.push(entry)
+            acknowledgements[idx] = { signature: entry.signature, pending: false }
+          }
+        }
+
+        const timestamp = BigInt(Date.now())
+        const salt = 0n
+        const bitset = Buffer.allocUnsafe(3)
+        bitset[0] = ackSet & 0xFF
+        bitset[1] = (ackSet >> 8) & 0xFF
+        bitset[2] = (ackSet >> 16) & 0xFF
+
+        client.write('chat_message', {
+          message,
+          timestamp,
+          salt,
+          signature: client.signMessage(message, timestamp, salt, toSign.map(ack => ack.signature)),
+          offset,
+          acknowledged: bitset
+        })
+      } else {
+        const timestamp = BigInt(Date.now())
+        client.write('chat_message', {
+          message,
+          timestamp,
+          salt: 0n,
+          signature: client.signMessage(message, timestamp)
+        })
+      }
+    } else client.write('chat', { message })
+  }
+
+  function fetchPreview (message) {
+    return new Promise((resolve, reject) => {
+      if (client.pendingPreview) {
+        if (options.queueMessages) queueMessagePreview(message, resolve, reject)
+        else reject(Error('Message could not be sent. Preview already in progress'))
+      } else {
+        requestPreview(message, resolve, reject)
+      }
+    })
+  }
+
+  function requestPreview (message, resolve, reject) {
+    client.previewId = client.previewId || 0
+    const id = client.previewId++
+    client.pendingPreview = { id, resolve, reject }
+
+    if (!client.lastPreviewRequest) client.lastPreviewRequest = 0
+    const diff = Date.now() - client.lastPreviewRequest
+
+    function sendPacket () {
+      client.write('chat_preview', {
+        query: id,
+        message
+      })
+
+      if (options.timeoutPreview) {
+        client.previewTimeout = setTimeout(() => {
+          throw Error('Server did not respond with preview in time') // Probably should not throw
+        }, options.timeoutPreview)
+      }
+    }
+
+    if (diff < 100) {
+      setTimeout(sendPacket, diff + 1)
+    } else sendPacket()
+  }
+
+  function queueMessagePreview (message, resolve, reject) {
+    if (!client.previewQueue) client.previewQueue = []
+    client.previewQueue.push({ message, resolve, reject })
+  }
+
+  // Packet handling
+
+  client.on('chat_preview', handleChatPreview)
+
+  client.on('player_info', handlePlayerInfo)
+
+  client.on('player_remove', handlePlayerRemove)
+
+  client.on('player_chat', handlePlayerChat)
+
+  client.on('system_chat', handleSystemChat)
+
+  client.on('hide_message', handleHideChat)
+
+  client.on('message_header', handleChatHeader)
+
+  client.on('should_display_chat_preview', packet => {
+    client.useChatPreview = packet.should_display_chat_preview
+  })
+
+  client.on('server_data', packet => {
+    client.useChatPreview = packet.previewsChat
+    client.signedChat = packet.enforcesSecureChat
```

Comment:
enforcesSecureChat does not exist on 1.19.2, it's secure profile

Reply:
actually looks like it was renamed in 1.19.1 to secureChat. Same thing per https://github.dev/extremeheat/extracted_minecraft_data/blob/client1.19.1/client/net/minecraft/server/MinecraftServer.java#L668